pgsql no longer compiles

pgsql no longer compiles against HEAD of hhvm master.

Here's the 1st handful of lines:

/home/rrh/dev/hphp-pgsql/pdo_pgsql.cpp:7:17: error: cannot declare variable ‘HPHP::s_pgsql_driver’ to be of abstract type ‘HPHP::PDOPgSql’
 static PDOPgSql s_pgsql_driver;
                 ^
In file included from /home/rrh/dev/hphp-pgsql/pdo_pgsql.cpp:1:0:
/home/rrh/dev/hphp-pgsql/pdo_pgsql.h:8:7: note:   because the following virtual functions are pure within ‘HPHP::PDOPgSql’:
 class PDOPgSql : public PDODriver {
       ^
In file included from /home/rrh/dev/hphp-pgsql/pdo_pgsql.h:5:0,
                 from /home/rrh/dev/hphp-pgsql/pdo_pgsql.cpp:1:
/usr/local/include/hphp/runtime/ext/pdo_driver.h:235:24: note: 	virtual HPHP::PDOResource* HPHP::PDODriver::createResourceImpl()
   virtual PDOResource* createResourceImpl() = 0;

--- a/pdo_pgsql.cpp
+++ b/pdo_pgsql.cpp
@@ -1,5 +1,6 @@
 #include "pdo_pgsql.h"
 #include "pdo_pgsql_connection.h"
+#include "pdo_pgsql_resource.h"
 
 namespace HPHP {
 
@@ -9,9 +10,10 @@ static PDOPgSql s_pgsql_driver;
 PDOPgSql::PDOPgSql() : PDODriver("pgsql") {
 }
 
-PDOConnection *PDOPgSql::createConnectionObject() {
-    return new PDOPgSqlConnection();
+PDOResource* PDOPgSql::createResourceImpl() {
+        return newres<PDOPgSqlResource>(std::make_shared<PDOPgSqlConnection>());
 }
+
 
 long pdo_attr_lval(const Array& options, int opt, long defaultValue){
     if(options.exists(opt)){

--- a/pdo_pgsql.h
+++ b/pdo_pgsql.h
@@ -5,11 +5,9 @@
 #include "hphp/runtime/ext/pdo_driver.h"
 
 namespace HPHP {
-class PDOPgSql : public PDODriver {
-public:
+struct PDOPgSql : public PDODriver {
     PDOPgSql();
-
-    virtual PDOConnection* createConnectionObject();
+    PDOResource* createResourceImpl() override;
 };
 
 long pdo_attr_lval(const Array& options, int opt, long defaultValue);

--- a/pdo_pgsql_connection.cpp
+++ b/pdo_pgsql_connection.cpp
@@ -1,6 +1,7 @@
 #include "pgsql.h"
 #include "pdo_pgsql_connection.h"
 #include "pdo_pgsql_statement.h"
+#include "pdo_pgsql_resource.h"
 #include "pdo_pgsql.h"
 #include "hphp/runtime/ext/stream/ext_stream.h"
 #include "hphp/runtime/vm/jit/translator-inline.h"
@@ -97,12 +98,12 @@ static int php_pdo_parse_data_source(const char *data_source,
 
 namespace HPHP {
 
-    PDOPgSqlConnection::PDOPgSqlConnection() : m_conn(nullptr), pgoid(InvalidOid) {
+    PDOPgSqlConnection::PDOPgSqlConnection() : m_server(nullptr), pgoid(InvalidOid) {
     }
 
     PDOPgSqlConnection::~PDOPgSqlConnection(){
-        if(m_conn){
-            delete m_conn;
+        if(m_server){
+            delete m_server;
         }
     }
 
@@ -143,20 +144,20 @@ namespace HPHP {
         conninfo << username << "'";
         conninfo << " connect_timeout=" << connect_timeout;
 
-        m_conn = new PQ::Connection(conninfo.str());
+        m_server = new PQ::Connection(conninfo.str());
 
-        if(m_conn->status() == CONNECTION_OK){
+        if(m_server->status() == CONNECTION_OK){
             return true;
         } else {
-            handleError(NULL, PHP_PDO_PGSQL_CONNECTION_FAILURE_SQLSTATE, m_conn->errorMessage());
+            handleError(NULL, PHP_PDO_PGSQL_CONNECTION_FAILURE_SQLSTATE, m_server->errorMessage());
             return false;
         }
     }
 
     bool PDOPgSqlConnection::closer(){
-        if(m_conn){
-            delete m_conn;
-            m_conn = NULL;
+        if(m_server){
+            delete m_server;
+            m_server = NULL;
         }
 
         return false;
@@ -167,7 +168,7 @@ namespace HPHP {
 
         const char* query = sql.data();
 
-        PQ::Result res = m_conn->exec(query);
+        PQ::Result res = m_server->exec(query);
 
         if(!res){
             // I think this error should be handled in a different way perhaps?
@@ -196,7 +197,7 @@ namespace HPHP {
     bool PDOPgSqlConnection::transactionCommand(const char* command){
         testConnection();
 
-        PQ::Result res = m_conn->exec(command);
+        PQ::Result res = m_server->exec(command);
 
         if(!res){
             // I think this error should be handled in a different way perhaps?
@@ -227,17 +228,17 @@ namespace HPHP {
     }
 
     void PDOPgSqlConnection::testConnection(){
-        if(!m_conn){
+        if(!m_server){
             handleError(NULL, "08003", NULL);
         }
     }
 
     bool PDOPgSqlConnection::checkLiveness(){
-        if(!m_conn){
+        if(!m_server){
             return false;
         }
 
-        return m_conn->status() == CONNECTION_OK;
+        return m_server->status() == CONNECTION_OK;
     }
 
     const char* PDOPgSqlConnection::sqlstate(PQ::Result& result){
@@ -247,7 +248,7 @@ namespace HPHP {
     bool PDOPgSqlConnection::quoter(const String& input, String &quoted, PDOParamType paramtype){
         switch(paramtype){
             case PDO_PARAM_LOB:
-                quoted = m_conn->escapeByteA(input.data(), input.length());
+                quoted = m_server->escapeByteA(input.data(), input.length());
                 return true;
                 break;
             default:
@@ -256,7 +257,7 @@ namespace HPHP {
 
                 buffer[0] = '\'';
                 int error;
-                size_t written = m_conn->escapeString(buffer.get()+1, input.c_str(), input.length()*2+1, &error);
+                size_t written = m_server->escapeString(buffer.get()+1, input.c_str(), input.length()*2+1, &error);
                 if(error){
                     return false;
                 }
@@ -282,7 +283,7 @@ namespace HPHP {
         } else {
             const char *values[1];
             values[0] = name;
-            PQ::Result res = m_conn->exec("SELECT CURRVAL($1)", 1, values);
+            PQ::Result res = m_server->exec("SELECT CURRVAL($1)", 1, values);
 
             ExecStatusType status = res.status();
 
@@ -301,17 +302,17 @@ namespace HPHP {
                 value = String(PG_VERSION);
                 break;
             case PDO_ATTR_SERVER_VERSION:
-                if(m_conn->protocolVersion() >= 3){ // Postgres 7.4 or later
-                    value = String(m_conn->parameterStatus("server_version"), CopyString);
+                if(m_server->protocolVersion() >= 3){ // Postgres 7.4 or later
+                    value = String(m_server->parameterStatus("server_version"), CopyString);
                 } else {
-                    PQ::Result res = m_conn->exec("SELECT VERSION()");
+                    PQ::Result res = m_server->exec("SELECT VERSION()");
                     if(res && res.status() == PGRES_TUPLES_OK){
                         value = String((char *)res.getValue(0, 0), CopyString);
                     }
                 }
                 break;
             case PDO_ATTR_CONNECTION_STATUS:
-                switch(m_conn->status()){
+                switch(m_server->status()){
                     case CONNECTION_STARTED:
                         value = String("Waiting for connection to be made.", CopyString);
                         break;
@@ -344,12 +345,12 @@ namespace HPHP {
                 }
                 break;
             case PDO_ATTR_SERVER_INFO: {
-                int spid = m_conn->backendPID();
+                int spid = m_server->backendPID();
 
                 std::stringstream result;
-                result << "PID: " << spid << "; Client Encoding: " << m_conn->parameterStatus("client_encoding");
-                result << "; Is Superusser: " << m_conn->parameterStatus("is_superuser") << "; Session Authorization: ";
-                result << m_conn->parameterStatus("session_authorization") << "; Date Style: " << m_conn->parameterStatus("DateStyle");
+                result << "PID: " << spid << "; Client Encoding: " << m_server->parameterStatus("client_encoding");
+                result << "; Is Superusser: " << m_server->parameterStatus("is_superuser") << "; Session Authorization: ";
+                result << m_server->parameterStatus("session_authorization") << "; Date Style: " << m_server->parameterStatus("DateStyle");
 
                 value = String(result.str());
             }
@@ -359,21 +360,6 @@ namespace HPHP {
         }
 
         return 1;
-    }
-
-    bool PDOPgSqlConnection::fetchErr(PDOStatement* stmt, Array &info){
-        if(stmt == nullptr){
-            info.append(m_lastExec == InvalidOid ? Variant(Variant::NullInit()) : Variant(m_lastExec));
-            info.append(err_msg.empty() ? Variant(Variant::NullInit()) : Variant(err_msg));
-            return true;
-        } else {
-            PDOPgSqlStatement* s = static_cast<PDOPgSqlStatement*>(stmt);
-            auto status = s->m_result.status();
-            auto emsg = s->err_msg;
-            info.append(status == InvalidOid ? Variant(Variant::NullInit()) : Variant(status));
-            info.append(emsg.empty() ? Variant(Variant::NullInit()) : Variant(emsg));
-            return true;
-        }
     }
 
     bool PDOPgSqlConnection::setAttribute(int64_t attr, const Variant &value){
@@ -387,11 +373,15 @@ namespace HPHP {
     }
 
     bool PDOPgSqlConnection::support(SupportedMethod method){
-        return method != MethodPersistentShutdown;
+        return true;
     }
 
-    bool PDOPgSqlConnection::preparer(const String& sql, sp_PDOStatement* stmt, const Variant &options){
-        PDOPgSqlStatement* s = NEWRES(PDOPgSqlStatement)(this, m_conn);
+    bool PDOPgSqlConnection::preparer(const String& sql, sp_PDOStatement *stmt, const Variant& options) {
+        auto rsrc = newres<PDOPgSqlResource>(
+            std::dynamic_pointer_cast<PDOPgSqlConnection>(shared_from_this()));
+
+        auto s = newres<PDOPgSqlStatement>(rsrc, m_server);
+
         *stmt = s;
 
         if(s->create(sql, options.toArray())){

--- a/pdo_pgsql_connection.h
+++ b/pdo_pgsql_connection.h
@@ -6,8 +6,8 @@
 
 #define PHP_PDO_PGSQL_CONNECTION_FAILURE_SQLSTATE "08006"
 namespace HPHP {
-
     class PDOPgSqlStatement;
+
     class PDOPgSqlConnection : public PDOConnection {
         friend class PDOPgSqlStatement;
     public:
@@ -35,13 +35,12 @@ namespace HPHP {
         virtual int getAttribute(int64_t attr, Variant &value);
         virtual bool setAttribute(int64_t attr, const Variant &value);
 
-        virtual bool fetchErr(PDOStatement* stmt, Array &info);
-
         String pgsqlLOBCreate();
 
-        virtual bool preparer(const String& sql, sp_PDOStatement *stmt, const Variant &options);
+        bool preparer(const String& sql, sp_PDOStatement *stmt, const Variant& options) override;
+
     private:
-        PQ::Connection* m_conn;
+        PQ::Connection* m_server;
         Oid pgoid;
         ExecStatusType m_lastExec;
         std::string err_msg;

--- a/pdo_pgsql_resource.h
+++ b/pdo_pgsql_resource.h
@@ -1,0 +1,24 @@
+#ifndef incl_HPHP_PDO_PGSQL_RESOURCE_H_
+#define incl_HPHP_PDO_PGSQL_RESOURCE_H_
+
+#include "hphp/runtime/ext/pdo_driver.h"
+#include "pdo_pgsql_connection.h"
+#include "pq.h"
+#include "stdarg.h"
+
+
+namespace HPHP {
+    class PDOPgSqlResource : public PDOResource {
+
+    public:
+        explicit PDOPgSqlResource(std::shared_ptr<PDOPgSqlConnection> conn)
+        : PDOResource(std::dynamic_pointer_cast<PDOConnection>(conn))
+        {}
+
+        std::shared_ptr<PDOPgSqlConnection> conn() const {
+            return std::dynamic_pointer_cast<PDOPgSqlConnection>(m_conn);
+        }
+    };
+}
+
+#endif

--- a/pdo_pgsql_statement.cpp
+++ b/pdo_pgsql_statement.cpp
@@ -1,20 +1,20 @@
+#include "pdo_pgsql_resource.h"
 #include "pdo_pgsql_statement.h"
 #include "pdo_pgsql_connection.h"
 #include "pdo_pgsql.h"
 #include "pgsql.h"
 #include <iomanip>
 
-
-#define STMT_HANDLE_ERROR(res) (*m_pdoconn).handleError(this, (*m_pdoconn).sqlstate(res), res.errorMessage())
+#define STMT_HANDLE_ERROR(res) (*m_conn).handleError(this, (*m_conn).sqlstate(res), res.errorMessage())
 
 namespace HPHP {
 
     unsigned long PDOPgSqlStatement::m_stmtNameCounter = 0;
     unsigned long PDOPgSqlStatement::m_cursorNameCounter = 0;
-    PDOPgSqlStatement::PDOPgSqlStatement(PDOPgSqlConnection* pdoconn, PQ::Connection* conn)
-        : m_pdoconn(pdoconn), m_conn(conn),
+    PDOPgSqlStatement::PDOPgSqlStatement(PDOPgSqlResource* conn, PQ::Connection* server)
+        : m_conn(conn->conn()), m_server(server),
           m_result(), m_isPrepared(false), m_current_row(0) {
-        this->dbh = pdoconn;
+        this->dbh = dynamic_cast<PDOResource*>(conn);
     }
 
     PDOPgSqlStatement::~PDOPgSqlStatement(){
@@ -26,7 +26,7 @@ namespace HPHP {
             if(m_isPrepared){
                 std::stringstream ss;
                 ss << "DEALLOCATE " << m_stmtName;
-                m_conn->exec(ss.str());
+                m_server->exec(ss.str());
             }
         }
 
@@ -34,9 +34,9 @@ namespace HPHP {
             // Do we need a check here maybe to see if we've actually got a cursor or not?
             std::stringstream ss;
             ss << "CLOSE " << m_cursorName;
-            m_conn->exec(ss.str());
+            m_server->exec(ss.str());
         }
-        m_pdoconn = NULL;
+        m_server = NULL;
         m_conn = NULL;
     }
 
@@ -50,13 +50,13 @@ namespace HPHP {
             // Disable prepared statements
             supports_placeholders = PDO_PLACEHOLDER_NONE;
         } else if (
-            m_pdoconn->m_emulate_prepare ||
+            m_conn->m_emulate_prepare ||
             (!options.empty() && pdo_attr_lval(options, PDO_ATTR_EMULATE_PREPARES, 0))
             ){
             supports_placeholders = PDO_PLACEHOLDER_NONE;
         }
 
-        if(supports_placeholders != PDO_PLACEHOLDER_NONE && m_conn->protocolVersion() > 2){
+        if(supports_placeholders != PDO_PLACEHOLDER_NONE && m_server->protocolVersion() > 2){
             named_rewrite_template = "$%d";
             String nsql;
             int ret = pdo_parse_params(this, sql, nsql);
@@ -64,8 +64,8 @@ namespace HPHP {
                 // Query was rewritten
             } else if (ret == -1){
                 // Query didn't parse - exception should have been thrown at this point
-                strncpy(m_pdoconn->error_code, error_code, 6);
-                m_pdoconn->error_code[5] = '\0';
+                strncpy(m_conn->error_code, error_code, 6);
+                m_conn->error_code[5] = '\0';
                 return false;
             } else {
                 // Original is great
@@ -91,13 +91,13 @@ namespace HPHP {
             if(m_isPrepared){
                 std::stringstream ss;
                 ss << "CLOSE " << m_cursorName;
-                m_result = m_conn->exec(ss.str());
+                m_result = m_server->exec(ss.str());
             }
 
             std::stringstream q;
             q << "DECLARE " << m_cursorName << " SCROLL CURSOR WITH HOLD FOR " << active_query_string.data();
 
-            m_result = m_conn->exec(q.str());
+            m_result = m_server->exec(q.str());
 
             status = m_result.status();
 
@@ -112,11 +112,11 @@ namespace HPHP {
 
             // Fetch to be able to get total number of rows
             q << "FETCH FORWARD 0 FROM " << m_cursorName;
-            m_result = m_conn->exec(q.str());
+            m_result = m_server->exec(q.str());
         } else if(m_stmtName.size() > 0) {
             if(!m_isPrepared){
 stmt_retry:
-                m_result = m_conn->prepare(m_stmtName.c_str(), m_resolvedQuery.c_str(), bound_params.size(), param_types.data());
+                m_result = m_server->prepare(m_stmtName.c_str(), m_resolvedQuery.c_str(), bound_params.size(), param_types.data());
 
                 status = m_result.status();
                 switch(status) {
@@ -128,10 +128,10 @@ stmt_retry:
                     default:
                         // Read Zend implementation for this one. I am not sure if this applies to hhvm as well or not
                         // but figure better leave it in here than not
-                        if(!strcmp(m_pdoconn->sqlstate(m_result), "42P05")){
+                        if(!strcmp(m_conn->sqlstate(m_result), "42P05")){
                             std::stringstream q;
                             q << "DEALLOCATE " << m_stmtName;
-                            m_conn->exec(q.str());
+                            m_server->exec(q.str());
                             goto stmt_retry;
                         } else {
                             STMT_HANDLE_ERROR(m_result);
@@ -152,13 +152,13 @@ stmt_retry:
             }
 
             if(params.size() != bound_params.size()){
-                m_pdoconn->handleError(this, "XX000", "Parameters not being bound correctly");
+                m_conn->handleError(this, "XX000", "Parameters not being bound correctly");
                 return false;
             }
 
-            m_result = m_conn->execPrepared(m_stmtName.c_str(), bound_params.size(), params.data(), param_lengths.data(), param_formats.data());
+            m_result = m_server->execPrepared(m_stmtName.c_str(), bound_params.size(), params.data(), param_lengths.data(), param_formats.data());
         } else {
-            m_result = m_conn->exec(active_query_string.data());
+            m_result = m_server->exec(active_query_string.data());
         }
 
         status = m_result.status();
@@ -177,7 +177,7 @@ stmt_retry:
 
         if(status == PGRES_COMMAND_OK){
             row_count = m_result.lcmdTuples();
-            m_pdoconn->pgoid = m_result.oidValue();
+            m_conn->pgoid = m_result.oidValue();
         } else {
             row_count = m_result.numTuples();
         }
@@ -248,7 +248,7 @@ stmt_retry:
             }
             std::string q = strprintf("FETCH %s FROM %s", oriStr.c_str(), m_cursorName.c_str());
 
-            m_result = m_conn->exec(q);
+            m_result = m_server->exec(q);
             ExecStatusType status = m_result.status();
 
             if(status != PGRES_COMMAND_OK && status != PGRES_TUPLES_OK){
@@ -288,7 +288,7 @@ stmt_retry:
 
         std::string q = strprintf("SELECT TYPNAME FROM PG_TYPE WHERE OID=%u", coltype);
 
-        PQ::Result res = m_conn->exec(q);
+        PQ::Result res = m_server->exec(q);
 
         ExecStatusType status = res.status();
 
@@ -370,7 +370,7 @@ stmt_retry:
                             if(bound_param_map.exists(param->name, true)){
                                 param->paramno = atoi(bound_param_map[param->name].asCStrRef().data() + 1)-1;
                             } else {
-                                m_pdoconn->handleError(this, "HY093", param->name.data());
+                                m_conn->handleError(this, "HY093", param->name.data());
                                 return false;
                             }
                         }
@@ -396,7 +396,7 @@ stmt_retry:
 
                     if(param->paramno >= 0){
                         if(param->paramno >= elems){
-                            m_pdoconn->handleError(this, "HY105", "Too many parameters bound");
+                            m_conn->handleError(this, "HY105", "Too many parameters bound");
                             return false;
                         }
 

--- a/pdo_pgsql_statement.h
+++ b/pdo_pgsql_statement.h
@@ -2,6 +2,7 @@
 #define incl_HPHP_PDO_PGSQL_STATEMENT_H_
 
 #include "hphp/runtime/ext/pdo_driver.h"
+#include "pdo_pgsql_resource.h"
 #include "pq.h"
 #include "stdarg.h"
 
@@ -14,12 +15,13 @@
 #define OIDOID      26
 
 namespace HPHP {
-    class PDOPgSqlConnection;
     class PDOPgSqlStatement : public PDOStatement {
-        friend class PDOPgSqlConnection;
+        friend PDOPgSqlConnection;
+
     public:
         DECLARE_RESOURCE_ALLOCATION(PDOPgSqlStatement);
-        PDOPgSqlStatement(PDOPgSqlConnection* m_pdoconn, PQ::Connection* pq);
+
+        PDOPgSqlStatement(PDOPgSqlResource* conn, PQ::Connection* server);
         virtual ~PDOPgSqlStatement();
 
         bool create(const String& sql, const Array &options);
@@ -37,9 +39,10 @@ namespace HPHP {
         virtual bool cursorCloser();
 
         virtual bool support(SupportedMethod method);
+
     private:
-        PDOPgSqlConnection* m_pdoconn;
-        PQ::Connection* m_conn;
+        std::shared_ptr<PDOPgSqlConnection> m_conn;
+        PQ::Connection* m_server;
         static unsigned long m_stmtNameCounter;
         static unsigned long m_cursorNameCounter;
         std::string m_stmtName;

